src/core/libmaven/PeakDetector.cpp

```diff
@@ -376,103 +527,435 @@ void PeakDetector::processSlices(vector<mzSlice*> &slices, string setName)
+        return;
+
+    // lambda: given a compound and its child indexes, clubs them with their
+    // most likely parent-group if possible, otherwise adds them to a ghost
+    auto makeMeta =
+        [this](Compound* compound,
+               vector<size_t> childIndexes,
+               unordered_map<Compound*, vector<size_t>> parentCompounds,
+               function<string(PeakGroup*)> nameFunc) {
+            auto& container = _mavenParameters->allgroups;
+            vector<size_t> orphans;
+            map<size_t, size_t> nonOrphans;
+            if (parentCompounds.count(compound) > 0) {
+                auto& parentIndexes = parentCompounds[compound];
+                auto result = _matchParentsToChildren(parentIndexes,
+                                                      childIndexes,
+                                                      container,
+                                                      nameFunc);
+                nonOrphans = result.first;
+                orphans = result.second;
             } else {
-                mz = compound->mz();
+                for (auto index : childIndexes)
+                    orphans.push_back(index);
             }
-            if (mzUtils::withinXMassCutoff(mz,
-                                           group.meanMz,
-                                           mavenParameters->massCutoffMerge)) {
-                PeakGroup groupWithTarget(group);
-                groupWithTarget.setCompound(compound);
 
-                // we should filter the annotated group based on its RT, if the
-                // user has restricted RT range
-                auto rtDiff = groupWithTarget.expectedRtDiff();
-                if (mavenParameters->identificationMatchRt
-                    && rtDiff > mavenParameters->identificationRtWindow) {
-                    continue;
-                }
+            unordered_map<size_t, vector<size_t>> metaGroups;
+            for (auto& elem : nonOrphans) {
+                size_t parentIndex = elem.second;
+                size_t childIndex = elem.first;
+                if (metaGroups.count(parentIndex) == 0)
+                    metaGroups[parentIndex] = {};
+                metaGroups[parentIndex].push_back(childIndex);
+            }
 
-                // since we are creating targeted groups, we should ensure they
-                // pass MS2 filtering criteria, if enabled
-                if (mavenParameters->matchFragmentationFlag
-                    && groupWithTarget.ms2EventCount > 0
-                    && groupFiltering.filterByMS2(groupWithTarget)) {
-                    continue;
-                }
+            if (!orphans.empty()) {
+                // for orphans, create a ghost, that will act as an empty parent
+                PeakGroup parentGroup(
+                    make_shared<MavenParameters>(*_mavenParameters),
+                    PeakGroup::IntegrationType::Ghost);
+                container.push_back(parentGroup);
```

Comment:
Note for self:
is there a new ghost parent for every orphan isotope or adduct? 
if there are M+1, M+2 for the same compound at multiple RTs, how is that handled? do they show up under different parents or the same?

Reply:
Same parent. Every labelled form for which there is no parent, is put under the one ghost parent. The ghost is only a container (for programming ease), and shows up as such in the UI (no data for any columns only a name and a group ID).